File: backend/utils/pincode.py
"""
utils/pincode.py — India Post PIN code API caller.
Resolves state, district, and taluka from a 6-digit Indian PIN code.
API: https://api.postalpincode.in/pincode/{pincode}
"""

import logging
import requests

logger = logging.getLogger(__name__)

PINCODE_API_BASE = "https://api.postalpincode.in/pincode"
REQUEST_TIMEOUT_S = 8


def resolve_pincode(pin_code: str) -> dict:
    """
    Calls the India Post PIN code API and extracts state, district, and taluka.

    Returns:
        {
            "state": "Maharashtra",
            "district": "Nashik",
            "taluka": "Nashik"   (Block/Taluka from PostOffice record)
        }

    Raises:
        ValueError: If the PIN code is invalid or the API returns no results.
        RuntimeError: If the API call fails (network error, timeout).
    """
    if not pin_code or len(pin_code) != 6 or not pin_code.isdigit():
        raise ValueError(f"Invalid PIN code format: '{pin_code}'. Must be exactly 6 digits.")

    url = f"{PINCODE_API_BASE}/{pin_code}"
    try:
        resp = requests.get(url, timeout=REQUEST_TIMEOUT_S)
        resp.raise_for_status()
        data = resp.json()
    except requests.RequestException as exc:
        logger.error("[Pincode] API call failed for %s: %s", pin_code, exc)
        raise RuntimeError(f"PIN code API unreachable: {exc}") from exc

    # API response structure: [{ "Status": "Success", "PostOffice": [...] }]
    if not data or data[0].get("Status") != "Success":
        raise ValueError(f"PIN code {pin_code} not found or invalid.")

    post_offices = data[0].get("PostOffice") or []
    if not post_offices:
        raise ValueError(f"No post office data found for PIN code {pin_code}.")

    # Use the first post office record for state/district/taluka
    po = post_offices[0]
    result = {
        "state":    po.get("State", ""),
        "district": po.get("District", ""),
        "taluka":   po.get("Block", ""),   # "Block" maps to Taluka in India Post API
    }
    logger.info("[Pincode] Resolved %s → %s", pin_code, result)
    return result

File: internal/service/pincode.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// PinCodeAPIBase is the India Post lookup endpoint.
const PinCodeAPIBase = "https://api.postalpincode.in/pincode"

// pincodeUserAgent is the K11 fix, approved by the owner.
//
// api.postalpincode.in TCP-resets the default `python-requests/x.y.z`
// User-Agent, which is why GET /farmer/pincode/<pin> returns 500 on every call
// in the Python backend and POST /farmer/location silently stores empty
// state/district/taluka. Go's default is `Go-http-client/1.1`, which is not
// blocked — but relying on that would be luck, so an explicit UA is set.
//
// This is the ONE sanctioned deviation from §0.8 (preserve known bugs): the
// owner accepted the working behaviour as the new baseline, so the two e15_*
// goldens (which captured the broken 500) are excluded from the parity gate.
const pincodeUserAgent = "MindstriX-Backend/2.0 (+https://github.com/SanTiwari07/NDVI_satellite)"

// PinCodeResult is the resolved administrative area.
type PinCodeResult struct {
	State    string `json:"state"`
	District string `json:"district"`
	Taluka   string `json:"taluka"`
}

// PinCodeError distinguishes "not found" (→ 404) from "unreachable" (→ 500),
// mirroring the ValueError / RuntimeError split in utils/pincode.py.
type PinCodeError struct {
	Msg      string
	NotFound bool
}

func (e *PinCodeError) Error() string { return e.Msg }

// PinCodeClient calls the India Post API.
type PinCodeClient struct {
	Base string
	HTTP *http.Client
}

// NewPinCodeClient returns a client with the Python's 4-second timeout.
func NewPinCodeClient() *PinCodeClient {
	return &PinCodeClient{
		Base: PinCodeAPIBase,
		HTTP: &http.Client{Timeout: 4 * time.Second},
	}
}

// Resolve looks up state / district / taluka for a 6-digit PIN.
func (c *PinCodeClient) Resolve(ctx context.Context, pin string) (*PinCodeResult, error) {
	if len(pin) != 6 || !allDigits(pin) {
		return nil, &PinCodeError{
			Msg:      fmt.Sprintf("Invalid PIN code format: '%s'. Must be exactly 6 digits.", pin),
			NotFound: true,
		}
	}

	base := c.Base
	if base == "" {
		base = PinCodeAPIBase
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet,
		base+"/"+url.PathEscape(pin), nil)
	if err != nil {
		return nil, &PinCodeError{Msg: "PIN code API unreachable: " + err.Error()}
	}
	req.Header.Set("User-Agent", pincodeUserAgent)
	req.Header.Set("Accept", "application/json")

	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, &PinCodeError{Msg: "PIN code API unreachable: " + err.Error()}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, &PinCodeError{
			Msg: fmt.Sprintf("PIN code API unreachable: status %d", resp.StatusCode),
		}
	}
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &PinCodeError{Msg: "PIN code API unreachable: " + err.Error()}
	}

	// The API answers with a single-element array.
	var payload []struct {
		Status     string `json:"Status"`
		PostOffice []struct {
			State    string `json:"State"`
			District string `json:"District"`
			Block    string `json:"Block"`
		} `json:"PostOffice"`
	}
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil, &PinCodeError{Msg: "PIN code API unreachable: " + err.Error()}
	}

	if len(payload) == 0 || payload[0].Status != "Success" {
		return nil, &PinCodeError{
			Msg:      fmt.Sprintf("PIN code %s not found or invalid.", pin),
			NotFound: true,
		}
	}
	if len(payload[0].PostOffice) == 0 {
		return nil, &PinCodeError{
			Msg:      fmt.Sprintf("No post office data found for PIN code %s.", pin),
			NotFound: true,
		}
	}

	po := payload[0].PostOffice[0]
	// "Block" is the India Post field that corresponds to Taluka.
	return &PinCodeResult{State: po.State, District: po.District, Taluka: po.Block}, nil
}

func allDigits(s string) bool {
	if s == "" {
		return false
	}
	return strings.IndexFunc(s, func(r rune) bool { return r < '0' || r > '9' }) < 0
}
